File: SRC/Services/PowerLab.UpdaterService/Worker.cs
using System.Diagnostics;
using Microsoft.Win32;
using PowerLab.Core.Contracts;
using PowerLab.Core.Models;
using PowerLab.Core.Models.Api;
using PowerLab.PluginContracts.Interfaces;

namespace PowerLab.UpdaterService;

public class Worker : BackgroundService
{
    private IAppLogger _logger;
    private IAppPackageRepository _packageRespository;
    private ILocalPluginRepository _localPluginRepository;
    private IRemotePluginRepository _remotePluginRepository;
    private readonly IServiceProvider _serviceProvider;
    private readonly string _basePath;
    private readonly string _appPackageCachePath;
    private readonly string _pluginPackageCachePath;
    private AppUpdateInfo _cachedAppUpdateInfo;
    private Dictionary<Guid, PluginUpdateInfo> _cachedPluginUpdateInfo = [];
    public Worker(IAppLogger logger, IServiceProvider serviceProvider)
    {
        _logger = logger;
        _serviceProvider = serviceProvider;

        _basePath = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData),
            "PowerLab", "UpdaterService", "Download");

        _appPackageCachePath = Path.Combine(_basePath, "PackageCache");
        _pluginPackageCachePath = Path.Combine(_basePath, "PluginCache");
        Directory.CreateDirectory(_basePath);
        Directory.CreateDirectory(_appPackageCachePath);
        Directory.CreateDirectory(_pluginPackageCachePath);
    }

    protected override async Task ExecuteAsync(CancellationToken stoppingToken)
    {
        _logger.Information("ExecuteAsync");

        using var scope = _serviceProvider.CreateScope();

        _packageRespository  = scope.ServiceProvider.GetRequiredService<IAppPackageRepository>();
        _localPluginRepository = scope.ServiceProvider.GetRequiredService<ILocalPluginRepository>();
        _remotePluginRepository = scope.ServiceProvider.GetRequiredService<IRemotePluginRepository>();

        while (!stoppingToken.IsCancellationRequested)
        {
            await Task.Delay(TimeSpan.FromSeconds(10), stoppingToken);

            await DownloadPackage();

            if (IsAppRunning()) continue;

            await InstallPackage();
        }
    }

    private async Task DownloadPackage()
    {
        await DownloadAppPackage();
        await DownloadPluginPackage();
    }

    private async Task DownloadPluginPackage()
    {
        _logger.Information("开始检测插件版本信息...");
        List<PluginRegistryModel> installedPlugins = _localPluginRepository.GetPluginRegistries();
        foreach (var plugin in installedPlugins)
        {
            var latestPackage = await _remotePluginRepository.GetLatestAsync(plugin.Manifest.Id);
            var latestVersion = latestPackage is null
                ? new Version("0.0.0")
                : new Version(latestPackage.Version);

            _logger.Information($"{plugin.Manifest.PluginName} 当前版本: {plugin.Manifest.Version}, 最新版本: {latestVersion}");

            var cachedVersion = _cachedPluginUpdateInfo.ContainsKey(plugin.Manifest.Id)
                ? new Version(_cachedPluginUpdateInfo[plugin.Manifest.Id].Version)
                : new Version("0.0.0");

            if (latestVersion <= new Version(plugin.Manifest.Version)) continue;

            if (latestVersion <= cachedVersion)
            {
                _logger.Information($"{plugin.Manifest.PluginName} {cachedVersion}已下载未安装，跳过下载");
                continue;
            }

            var targetPath = Path.Combine(_pluginPackageCachePath, latestPackage.FileName);
            bool result = await _remotePluginRepository.DownloadLatestAsync(plugin.Manifest.Id, "stable", targetPath, null);
            if (!result)
            {
                _logger.Information("插件安装包下载失败");
                continue;
            }
            _cachedPluginUpdateInfo[plugin.Manifest.Id] = new PluginUpdateInfo(plugin.Manifest.Id, latestPackage.Version, targetPath);
        }
    }

    private async Task DownloadAppPackage()
    {
        _logger.Information("开始检测客户端版本信息...");

        Version currentVersion = GetInstalledVersion();
        var newestVersion = await GetLastestVersionAsync();
        _logger.Information($"当前版本: {currentVersion}, 最新版本: {newestVersion.Version}");

        var newestVer = new Version(newestVersion.Version);
        var cachedVer = new Version(_cachedAppUpdateInfo?.Version ?? "0.0.0");
        if (newestVer <= currentVersion || newestVer <= cachedVer)
        {
            _logger.Information("未检测到新版本");
            return;
        }

        _logger.Information("正在下载新版本安装包");
        var targetPath = Path.Combine(_appPackageCachePath, newestVersion.FileName);
        var progress = new Progress<double>(p => { });
        bool result = await _packageRespository.DownloadLatestAsync("stable", targetPath, progress);
        _cachedAppUpdateInfo = new AppUpdateInfo(newestVersion.Version, targetPath);
    }

    private async Task InstallPackage()
    {
        await InstallPluginPackage();
        await InstallAppPackage();
    }

    private async Task InstallAppPackage()
    {
        if (_cachedAppUpdateInfo is null)
        {
            _logger.Information("没有新版本需要安装");
            return;
        }

        _logger.Information("开始启动客户端安装程序");
        var processStartInfo = new ProcessStartInfo
        {
            FileName = _cachedAppUpdateInfo.FilePath,
            Arguments = "/quiet",
            UseShellExecute = false,
            CreateNoWindow = true
        };
        using Process? process = Process.Start(processStartInfo);
        if (process is not null)
        {
            await process.WaitForExitAsync();
            _logger.Information("客户端安装程序执行完成，继续检测更新。");
            File.Delete(_cachedAppUpdateInfo.FilePath);
            _cachedAppUpdateInfo = null;
            return;
        }
        _logger.Information("客户端安装程序启动失败");
    }

    private async Task InstallPluginPackage()
    {
        var cachedPluginIdList = _cachedPluginUpdateInfo.Keys.ToList();
        string? installFolder = Path.GetDirectoryName(GetInstallPath());
        if (installFolder is null)
        {
            _logger.Information("找不到客户端的安装目录");
            return;
        }

        string pluginInstallerPath = Path.Combine(installFolder, "PluginInstaller.exe");
        foreach (var pluginId in cachedPluginIdList)
        {
            var pluginUpdateInfo = _cachedPluginUpdateInfo[pluginId];
            _logger.Information($"开始安装插件 {pluginId} 的新版本 {pluginUpdateInfo.Version}");
            var processStartInfo = new ProcessStartInfo
            {
                FileName = pluginInstallerPath,
                UseShellExecute = false,
                CreateNoWindow = true,
            };
            processStartInfo.ArgumentList.Add(pluginUpdateInfo.FilePath);
            processStartInfo.ArgumentList.Add("--nowindow");
            using Process? process = Process.Start(processStartInfo);
            if (process is not null)
            {
                await process.WaitForExitAsync();
                _logger.Information($"插件 {pluginId} 的安装程序执行完成。");
                File.Delete(pluginUpdateInfo.FilePath);
                _cachedPluginUpdateInfo.Remove(pluginId);
                continue;
            }
            _logger.Information($"插件 {pluginId} 的安装程序启动失败。");
        }
    }


    public override async Task StopAsync(CancellationToken cancellationToken)
    {
        _logger.Information("StopAsync");
        await base.StopAsync(cancellationToken);
    }
    public override async Task StartAsync(CancellationToken cancellationToken)
    {
        _logger.Information("StartAsync");
        await base.StartAsync(cancellationToken);
    }
    private static Version GetInstalledVersion()
    {
        const string keyPath = @"Software\PowerLab";
        using RegistryKey? key = Registry.LocalMachine.OpenSubKey(keyPath);
        if (key == null) return new Version("1.0.0");

        object? value = key.GetValue("InstallVersion");
        return new Version($"{value ?? "1.0.0"}");
    }
    private static string GetInstallPath()
    {
        const string keyPath = @"Software\PowerLab";
        using RegistryKey? key = Registry.LocalMachine.OpenSubKey(keyPath);
        if (key == null) return null;

        object? value = key.GetValue("InstallPath");
        return value?.ToString();
    }
    private async Task<AppVersionInfo> GetLastestVersionAsync()
    {
        var latestVersion = await _packageRespository.GetLatestAsync();
        return latestVersion ?? new AppVersionInfo { Version = "1.0.0" };
    }

    private static bool IsAppRunning()
    {
        List<Process> allProcesses =
        [.. Process.GetProcessesByName("PowerLab"),
         .. Process.GetProcessesByName("PlixInstaller")];

        if (allProcesses.Count <= 0) return false;

        string? installFolder = Path.GetDirectoryName(GetInstallPath());
        List<Process> runningProcesses =
            [.. allProcesses.Where(p => Path.GetDirectoryName(p.MainModule.FileName) == installFolder)];

        if (runningProcesses.Count <= 0) return false;

        return runningProcesses.Any(p => !p.HasExited);
    }
}

public record AppUpdateInfo(string Version, string FilePath);

public record PluginUpdateInfo(Guid PluginId, string Version, string FilePath);

File: SRC/Services/PowerLab.UpdaterService/Worker.fs
namespace PowerLab.UpdaterService

open System
open System.Threading
open System.Threading.Tasks
open Microsoft.Extensions.Hosting
open System.IO
open Microsoft.Extensions.DependencyInjection
open PowerLab.Core.Contracts
open PowerLab.PluginContracts.Interfaces
open Microsoft.Win32
open System.Diagnostics
open Microsoft.Data.Sqlite
open System.Text.Json

[<CLIMutable>]
type AppUpdateInfo = {
    Version: string
    FilePath: string
}

[<CLIMutable>]
type PluginUpdateInfo = {
    PluginId: Guid
    Version: string
    FilePath: string
}

[<CLIMutable>]
type PendingUpdate = {
    mutable App: AppUpdateInfo option
    mutable Plugins: Map<Guid, PluginUpdateInfo>
}

type Worker(logger: IAppLogger, serviceProvider: IServiceProvider) =
    inherit BackgroundService()
    
    let basePath = 
        Path.Combine(Environment.GetFolderPath Environment.SpecialFolder.CommonApplicationData, 
                     "PowerLab", "UpdaterService", "Download")

    let appPackageCachePath = Path.Combine(basePath, "PackageCache")
    let pluginPackageCachePath = Path.Combine(basePath, "PluginCache")
    let configFilePath = Path.Combine(basePath, "pending_updates.json")
    
    let mutable pendingUpdate = { App = None; Plugins = Map.empty }

    let saveState (state: PendingUpdate) =
        try
            let options = JsonSerializerOptions(WriteIndented = true)
            let json = JsonSerializer.Serialize(state, options)
            File.WriteAllText(configFilePath, json)
        with ex -> logger.Information $"保存状态文件失败: {ex.Message}"

    let loadState () =
        try
            if File.Exists configFilePath then
                let json = File.ReadAllText(configFilePath)
                pendingUpdate <- JsonSerializer.Deserialize<PendingUpdate>(json)
                logger.Information "成功从本地恢复待更新状态"
        with ex -> 
            logger.Information $"加载状态文件失败，将重新开始: {ex.Message}"
            pendingUpdate <- { App = None; Plugins = Map.empty }

    do
        [basePath; appPackageCachePath; pluginPackageCachePath]
        |> List.iter (fun path -> path |> Directory.CreateDirectory |> ignore)
        loadState()

    let getRegistryValue registryKey = 
        use key = Registry.LocalMachine.OpenSubKey @"Software\PowerLab"
        match key with
        | null -> None
        | _ -> key.GetValue registryKey |> Option.ofObj |> Option.map string

    let getInstallPath () = getRegistryValue "LauncherPath"

    let getInstalledVersion () =
        getRegistryValue "CurrentVersion"
        |> Option.defaultValue "1.0.0"
        |> Version

    let isAppRunning () =
        let installFolder = getInstallPath() |> Option.map Path.GetDirectoryName
        let processNames = ["PowerLab"; "PlixInstaller"]

        processNames
        |> List.collect (Process.GetProcessesByName >> List.ofArray)
        |> List.exists (fun p -> 
            try
                let pDir = Path.GetDirectoryName p.MainModule.FileName        
                installFolder = Some pDir && not p.HasExited    
            with _ -> false)

    let downloadAppPackage (packageRepo: IAppPackageRepository) = async {
        logger.Information "开始检测客户端版本信息..."
        let currentVersion = getInstalledVersion()
        let! latestInfo = packageRepo.GetLatestAsync() |> Async.AwaitTask
        let newestVersion = if isNull latestInfo then "1.0.0" else latestInfo.Version

        let newestVer = Version newestVersion
        let cachedVer = pendingUpdate.App |> Option.map (fun i -> Version i.Version) |> Option.defaultValue (Version "0.0.0")

        if newestVer > currentVersion && newestVer > cachedVer then
            logger.Information $"发现新版本 {newestVersion}，正在下载..."
            let targetPath = Path.Combine(appPackageCachePath, latestInfo.FileName)
            let! success = packageRepo.DownloadLatestAsync("stable", targetPath, Progress<double> ignore) |> Async.AwaitTask
            if success then
                pendingUpdate.App <- Some { Version = newestVersion; FilePath = targetPath }
                saveState pendingUpdate
                logger.Information "客户端下载完成并已保存状态"
        else
            logger.Information "未检测到更高版本的客户端"
    }

    let downloadPluginPackage (localRepo: ILocalPluginRepository, remoteRepo: IRemotePluginRepository) = async {
        logger.Information "开始检测插件版本信息..."
        let installedPlugins = localRepo.GetPluginRegistries() |> List.ofSeq
        SqliteConnection.ClearAllPools()

        for plugin in installedPlugins do
            let! latestPackage = remoteRepo.GetLatestAsync plugin.Manifest.Id |> Async.AwaitTask
            let latestVersion = if isNull latestPackage then Version "0.0.0" else Version latestPackage.Version
            
            let cachedVersion = 
                match pendingUpdate.Plugins.TryFind plugin.Manifest.Id with
                | Some info -> Version info.Version
                | None -> Version "0.0.0"

            if latestVersion > Version plugin.Manifest.Version && latestVersion > cachedVersion then
                let targetPath = Path.Combine(pluginPackageCachePath, latestPackage.FileName)
                let! result = remoteRepo.DownloadLatestAsync(plugin.Manifest.Id, "stable", targetPath, null) |> Async.AwaitTask
                if result then
                    let newPlugins = pendingUpdate.Plugins.Add(plugin.Manifest.Id, { PluginId = plugin.Manifest.Id; Version = latestPackage.Version; FilePath = targetPath})
                    pendingUpdate.Plugins <- newPlugins
                    saveState pendingUpdate
                    logger.Information $"插件 {plugin.Manifest.PluginName} 下载完成"
                else
                    logger.Information $"插件 {plugin.Manifest.PluginName} 下载失败"
    }

    let installAppPackage () = async {
        match pendingUpdate.App with
        | None -> ()
        | Some info ->

            logger.Information "开始启动客户端安装程序"
            let psi = ProcessStartInfo(info.FilePath, Arguments = "/quiet /log debug.log", UseShellExecute = false, CreateNoWindow = true)
            try
                use p = Process.Start psi  
                if not (isNull p) then
                    pendingUpdate.App <- None
                    saveState pendingUpdate

                    do! p.WaitForExitAsync() |> Async.AwaitTask
                    logger.Information "客户端更新完成"
                    if File.Exists info.FilePath then File.Delete info.FilePath
            with ex -> logger.Information $"安装客户端时发生异常: {ex.Message}"
    }

    let installPluginPackageCore (pluginIds: Guid list) installFolder = task {
        let pluginInstallerPath = Path.Combine(installFolder, "PluginInstaller.exe")
        
        for pluginId in pluginIds do
            match pendingUpdate.Plugins.TryFind pluginId with
            | Some info ->
                logger.Information $"正在更新插件: {pluginId}"
                let psi = ProcessStartInfo(pluginInstallerPath, UseShellExecute = false, CreateNoWindow = true)
                psi.ArgumentList.Add info.FilePath
                psi.ArgumentList.Add "--nowindow"

                try
                    use p = Process.Start psi
                    if not (isNull p) then
                        do! p.WaitForExitAsync() |> Async.AwaitTask
                        if File.Exists info.FilePath then File.Delete info.FilePath
                        pendingUpdate.Plugins <- pendingUpdate.Plugins.Remove pluginId
                        saveState pendingUpdate
                with ex -> logger.Information $"安装插件 {pluginId} 时异常: {ex.Message}"
            | None -> ()
    }

    let installPluginPackage () = async {
        let cachedPluginIdList = pendingUpdate.Plugins.Keys |> Seq.toList
        if List.isEmpty cachedPluginIdList then return ()

        let installPathOpt = getInstallPath() |> Option.map Path.GetDirectoryName
        match installPathOpt with
        | None -> logger.Information "找不到客户端的安装目录，无法安装插件"
        | Some installFolder -> do! installPluginPackageCore cachedPluginIdList installFolder |> Async.AwaitTask
    }

    override _.ExecuteAsync(stoppingToken: CancellationToken) = task {
        logger.Information "更新服务工作循环已启动"
        
        while not stoppingToken.IsCancellationRequested do
            try
                use scope = serviceProvider.CreateScope()
                let packageRepo = scope.ServiceProvider.GetRequiredService<IAppPackageRepository>()
                let localRepo = scope.ServiceProvider.GetRequiredService<ILocalPluginRepository>()
                let remoteRepo = scope.ServiceProvider.GetRequiredService<IRemotePluginRepository>()

                do! downloadAppPackage packageRepo |> Async.StartAsTask
                do! downloadPluginPackage (localRepo, remoteRepo) |> Async.StartAsTask

                if not (isAppRunning()) then
                    do! installPluginPackage () |> Async.StartAsTask
                    do! installAppPackage () |> Async.StartAsTask
                else
                    if pendingUpdate.App.IsSome || not pendingUpdate.Plugins.IsEmpty then
                        logger.Information "检测到待更新项，但程序正在运行，等待退出后安装"

                do! Task.Delay(TimeSpan.FromMinutes 1.0, stoppingToken)

            with ex ->
                logger.Information $"工作循环中发生错误: {ex.Message}"
    }

    override _.StartAsync ct = 
        logger.Information "Updater Service Starting..."
        base.StartAsync ct

    override _.StopAsync ct = 
        logger.Information "Updater Service Stopping..."
        base.StopAsync ct